cmd/stellar-rpc/internal/methods/get_ledgers.go

```diff
@@ -159,99 +161,104 @@ func (h ledgersHandler) parseCursor(cursor string, ledgerRange protocol.LedgerSe
 //  1. Entire range is available in local db.
 //  2. Entire range is unavailable in local db so fetch fully from datastore.
 //  3. Range partially available in the local db with the rest fetched from the datastore.
-func (h ledgersHandler) fetchLedgers(ctx context.Context, start uint32,
-	end uint32, format string, readTx db.LedgerReaderTx, localLedgerRange protocol.LedgerSeqRange,
+func (h ledgersHandler) fetchLedgers(
+	ctx context.Context,
+	start, end uint32, format string,
+	readTx db.LedgerReaderTx,
+	localLedgerRange protocol.LedgerSeqRange,
 ) ([]protocol.LedgerInfo, error) {
-	fetchFromLocalDB := func(start uint32, end uint32) ([]xdr.LedgerCloseMeta, error) {
+	limit := end - start + 1
+	result := make([]protocol.LedgerInfo, 0, limit)
+
+	addToResult := func(ledgers []db.LedgerMetadataChunk, metadata []xdr.LedgerCloseMeta) error {
+		// First, convert deserialized structures into serialized ones, if present.
+		chunks, err := metaToChunk(metadata)
+		if err != nil {
+			return &jrpc2.Error{
+				Code:    jrpc2.InternalError,
+				Message: fmt.Sprintf("error serializing ledgers: %v", err),
+			}
+		}
+		ledgers = append(ledgers, chunks...)
+
+		// Then, transform them all into JSON responses.
+		for _, chunk := range ledgers {
+			if len(result) >= int(limit) {
+				break
+			}
+
+			info, err := parseLedgerInfo(chunk, format)
+			if err != nil {
+				return &jrpc2.Error{
+					Code: jrpc2.InternalError,
+					Message: fmt.Sprintf("error processing ledger %d: %v",
+						chunk.Header.Header.LedgerSeq, err),
+				}
+			}
+			result = append(result, info)
+		}
+		return nil
+	}
+
+	fetchFromLocalDB := func(start, end uint32) error {
 		ledgers, err := readTx.BatchGetLedgers(ctx, start, end)
 		if err != nil {
-			return nil, &jrpc2.Error{
+			return &jrpc2.Error{
 				Code:    jrpc2.InternalError,
 				Message: fmt.Sprintf("error fetching ledgers from db: %v", err),
 			}
 		}
-		return ledgers, nil
+
+		return addToResult(ledgers, []xdr.LedgerCloseMeta{})
```

Comment:
nit, could pass nil instead of []xdr.LedgerCloseMeta{}, same for the empty slice allocation on line 230.